sys/src/9/port/sysfile.c

```diff
@@ -742,21 +741,24 @@ read(int fd, void *p, int32_t n, int64_t off)
 	 * The number of bytes read on this fd (c->offset) may be different
 	 * due to rewritings in mountfix.
 	 */
+	if(ispread){	/* use and maintain channel's offset */
```

Comment:
I think you can simplify this to,

    if(sequential || off == -1LL) {
        off = c->offset;
        sequential = 1;
    }